### PR TITLE
Allow use of `ForwardDiff` and `ReverseDiff` to calculate gradient of augmented ELBO

### LINKED ISSUE #76: `aux_posterior` is not AD-ready
I tried to AD `aug_elbo` in the `NegBinomialLikelihood` example, i.e. (removed unnecessary bits), purposefully avoiding ParameterHandling.jl and trying only with `ForwardDiff.gradient`

```julia
# # Negative Binomial

# We load all the necessary packages
using AbstractGPs
using ApproximateGPs
using AugmentedGPLikelihoods
using Distributions
using ForwardDiff # <-- try this first
using LinearAlgebra

# We create some random data (sorted for plotting reasons)
N = 100
x = range(-10, 10; length=N)
kernel = with_lengthscale(SqExponentialKernel(), 2.0)
gp = GP(kernel)
lik = NegBinomialLikelihood(15)
lf = LatentGP(gp, lik, 1e-6)
f, y = rand(lf(x));

# ## ELBO
# How can one compute the Augmented ELBO?
# Again AugmentedGPLikelihoods provides helper functions
# to not have to compute everything yourself
function aug_elbo(lik, u_post, x, y)
    qf = marginals(u_post(x))
    qΩ = aux_posterior(lik, y, qf)
    return expected_logtilt(lik, qΩ, y, qf) - aux_kldivergence(lik, qΩ, y) -
           kldivergence(u_post.approx.q, u_post.approx.fz)     # approx.fz is the prior and approx.q is the posterior 
end

function u_posterior(fz, m, S)
    return posterior(SparseVariationalApproximation(Centered(), fz, MvNormal(m, S)))
end

# ## Try to differentiate loss function

function makeloss(x, y)
    N = length(x)
    function loss(θ)
        k = ScaledKernel(
            RBFKernel() ∘ ScaleTransform(inv(θ.lengthscale)), 
            θ.variance
        )
        gp = GP(k)
        lik = NegBinomialLikelihood(θ.r)
        fz = gp(x, 1e-8);
        u_post = u_posterior(fz, zeros(N), Matrix{Float64}(I(N)))
        return aug_elbo(lik, u_post, x, y)
    end
end

θ = (
    variance = 1.,
    lengthscale = 1.,
    r = 15.
)

θ_flat = vcat(
    θ.variance, 
    θ.lengthscale,
    θ.r
)

function vec_to_nt(v)
    return (
        variance = v[1],
        lengthscale = v[2],
        r = v[3]
    )
end

vec_to_nt(θ_flat) == θ

loss = makeloss(x, y)
loss(θ) # works!
ForwardDiff.gradient(loss ∘ vec_to_nt, θ_flat) # MethodError
```

There is an easy fix (happy to open a PR): change the definition of `aux_posterior` as
```julia
function aux_posterior(lik::AbstractLikelihood, y, f)
    v = sqrt.(second_moment.(f))
    return For(TupleVector(; y=y, c=v)) do φ
        NTDist(PolyaGamma(φ.y + lik.r, φ.c)) # Distributions uses a different parametrization
    end
    return aux_posterior!(init_aux_posterior(lik, length(y)), lik, y, f)
end
```

```julia
julia> ForwardDiff.gradient(loss ∘ vec_to_nt, θ_flat)
3-element Vector{Float64}:
  5.790557942012172e7
 -1.9761748845444782e9
 16.184871970106013
```

BTW: is it expected that the values of the augmented ELBO are so much larger in magnitude than the normal ELBO?

### PR DESCRIPTION
**Pull request by simsurace**
The Poisson case does not work yet because I haven't yet managed to figure out how to make it ignore the derivative of `LogExpFunctions._logistic_bounds`, which depends only on the type and is therefore constant. One way to solve this would be to add explicit overloads for `ForwardDiff.Dual` etc., but this adds dependencies.

I had to remove a few type restrictions, while they shouldn't break anything they offer less protection for users.

After the Poisson case is fixed, this closes #76.

### CODE REVIEW

**Comment by simsurace** [comment]
This is not required per https://github.com/JuliaGaussianProcesses/AugmentedGPLikelihoods.jl/issues/76#issuecomment-1085551285
